main/src/main/java/cgeo/geocaching/utils/OfflineTranslateUtils.java

```diff
@@ -127,21 +132,38 @@ public static void getTranslator(final Activity activity, final Language sourceL
     }
 
     public static void translateParagraph(final Translator translator, final OfflineTranslateUtils.Status status, final String text, final Consumer<String> consumer, final Consumer<Exception> errorConsumer) {
-        final List<String> origText = Arrays.asList(text.split("\n"));
-        final List<String> translatedText = new ArrayList<>();
-        for (String p : origText) {
-            translator.translate(p)
-                    .addOnSuccessListener(translation -> {
-                        translatedText.add(translation);
-                        if (origText.size() == translatedText.size()) {
-                            consumer.accept(String.join("\n", translatedText));
-                            status.updateProgress();
-                        }
-                    })
-                    .addOnFailureListener(e -> {
-                        status.abortTranslation();
-                        errorConsumer.accept(e);
-                    });
+        final Document document = Jsoup.parse(text);
+        final Elements elements = document.children().select("*");
+        final AtomicInteger remaining = new AtomicInteger(elements.size());
+        for (Element element : elements) {
+            final String current = element.ownText();
+            if (StringUtils.isBlank(current)) {
+                remaining.decrementAndGet();
+            } else {
+                // translate
+                final String[] origText = current.split("\n");
+                final AtomicInteger remaining2 = new AtomicInteger(origText.length);
+                for (int i = 0, max = origText.length; i < max; i++) {
+                    final int j = i;
+                    translator.translate(origText[i])
+                        .addOnSuccessListener(translation -> {
+                            origText[j] = translation;
+                            if (remaining2.decrementAndGet() == 0) {
+                                element.text(String.join("\n", origText));
```

Comment:
...but you set it again using `text()`. Could that build up a difference in resulting html?

Reply:
There is no `setOwnText()` or something like that - but as `text()` belongs to a single element, it just sets that's (own) text.